packages/core/src/interpreter.ts

```diff
@@ -477,67 +402,70 @@ export class Interpreter<
         });
       } else {
         this.stop();
+
+        // TODO: improve this
         throw event.data.data;
       }
     }
 
-    const nextState = this._nextState(event);
+    const nextState = this.behavior.transition(
+      this._state,
+      event,
+      this._actorContext
+    );
 
     this.update(nextState);
 
-    if (errored) {
+    if (event.name === 'xstate.stop') {
+      this._stop();
+    } else if (errored) {
       this.stop();
     }
   }
 
-  private _stop() {
-    this.listeners.clear();
+  /**
+   * Stops the interpreter and unsubscribe all listeners.
+   *
+   * This will also notify the `onStop` listeners.
+   */
+  public stop(): this {
+    delete this.__initial;
+    // const mailbox = this.mailbox;
+
+    // this._stop();
```

Comment:
It feels like there is an important logic in this function that should be called whenever public `.stop` gets called.